pageserver/src/pgdatadir_mapping.rs

```diff
@@ -196,6 +198,45 @@ impl<R: Repository> DatadirTimeline<R> {
         Ok(exists)
     }
 
+    pub fn lookup_lsn(
+        &self,
+        timestamp_opt: Option<TimestampTz>,
+        xid_opt: Option<TransactionId>,
+    ) -> Result<Option<(ZTimelineId, Lsn)>> {
+        let lsn = self.tline.get_last_record_lsn();
+        let mut segments: Vec<u32> = self
+            .list_slru_segments(SlruKind::Clog, lsn)?
+            .into_iter()
+            .collect();
+        segments.sort_unstable();
+        for &segno in segments.iter().rev() {
+            let nblocks = self.get_slru_segment_size(SlruKind::Clog, segno, lsn)?;
+            for blknum in (0..nblocks).rev() {
```

Comment:
What we need is to iterate through all WAL records associated with CLOG pages.
The most efficient way is definitely find out layers containing this wal records and traverse them in LSN decreasing order. But is seems to be non trivial, taken in account that layers can overlap, contains multiple relish and there are branches. So I prefer to be as close to get_reconstrucrt_data as possible.
Efficiency seems to be not so critical here, because in most cases user will want to go not so far in the past.
So we really first do loop through all CLOG pages and search WAL records associated with this pages.

Reply:
I think this will not work correctly, if some CLOG segments have been truncated away. The LSN you're looking for might be in a WAL record for a CLOG segment / page that doesn't exist at the latest LSN.